qmi/core/transport.py

```diff
@@ -1019,31 +1020,53 @@ def write(self, data: bytes) -> None:
         except vxi11.vxi11.Vxi11Exception as err:
             raise QMI_InstrumentException() from err
 
-    def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+    def read(self, nbytes: int, timeout: Optional[float] = None) -> bytes:
         self._check_is_open()
+        nbuf = len(self._read_buffer)
+        if nbuf >= nbytes:
+            # The requested number of bytes are already in the buffer. Return them immediately.
+            ret = self._read_buffer[:nbytes]
+            self._read_buffer = self._read_buffer[nbytes:]
+            return ret
+
         old_timeout = self._safe_instr.timeout
         if timeout:
             self._safe_instr.timeout = timeout
+
         try:
-            data = self._safe_instr.read_raw(nbytes)
+            while len(self._read_buffer) < nbytes:
+                nbytes_batch = min(512, nbytes - len(self._read_buffer))
```

Comment:
minor: can't we replace 512 with `self._safe_instr.max_recv_size`. Then we stick to what the driver can handle?

Reply:
Yes, that sounds like a good idea!